### PR TITLE
Feature/spec errors

### PR DESCRIPTION
**Pull request by shabbyrobe**
Hello again! Thank you for inviting me to collaborate. I have really enjoyed working on the code here so far, it has been very easy to work with.

This PR is another structural one, and once again it is a somewhat deep change. It adds support for backends to return rich errors that reflect the XML structure S3 returns. I've used a few tricks to pull that off; the errors returned are actually the same serialisable values the XML marshaller will return to the client!

You can also, in a pinch, just return ErrorCode directly as the HTTP handler will turn it into an XML response just the same. It's this design decision I'm a bit less certain of but it was convenient enough in practice and helped with the refactoring of the existing backends. I am just a bit more concerned it might give Backend authors license to leave out too much information. Maybe it can evolve over time to the helper constructor style, like `BucketNotFound()` and friends, then we can remove this quirk of ErrorCodes. What do you think?

There are a lot of extra errors implemented here than the core ones that the backends needed to support the basic functionality. I did this in order to try to put the design decisions I had made to the test to make sure I had laid enough groundwork to cover future changes. Things like "RequestTimeTooSkewed" really didn't need to be implemented from a usability point of view (I didn't even realise that was a thing AWS would fail on!), but it was very useful as a driver to help get the shape and behaviour of the derived response types right (and it can't hurt to have a few extra checks around!) You can see from the commit history that it took me a few goes, and that doesn't include the failed attempts that never got committed in the first place 😆

### CODE REVIEW

**Comment by shabbyrobe** [comment]
Hi @johannesboyne, happy new year! Sorry to pester you, but I was wondering if you had had a chance to look over this yet? I have a few other PRs almost ready to submit but this one is a dependency.

File: gofakes3.go [hunk 1]
```diff
@@ -419,3 +525,53 @@ func formatHeaderTime(t time.Time) string {
...
+//
+// 1. Bucket names must comply with DNS naming conventions.
+// 2. Bucket names must be at least 3 and no more than 63 characters long.
+// 3. Bucket names must not contain uppercase characters or underscores.
+// 4. Bucket names must start with a lowercase letter or number.
+//
+// The DNS RFC confirms that the valid range of characters in an LDH label is 'a-z0-9-':
+// https://tools.ietf.org/html/rfc5890#section-2.3.1
+//
+func ValidateBucketName(name string) error {
```
Comment (johannesboyne) [inline gofakes3.go]: One could consider adding a test for this to codify not only the rules but as well the correct implementation of them.

Comment (shabbyrobe) [reply gofakes3.go]: Definitely! Done.

**Comment by shabbyrobe** [comment]
Thanks heaps for the review!

I've fixed the commend and added the test. I agree that `gofakes3.go` was getting a bit big - I really wanted to start breaking it down into separate pieces too. I was saving that for a later PR, but there's no time like the present! I've just decomposed into separate files for now, which helps a bit; as you say, we can clean this up further in future improvements.

**Comment by johannesboyne** [comment]
Very cool! Looking forward for the next improvements and thank you for making this project better!

**Pull request merged**